File: 算法设计与分析/ch3分治法/Quick_Sort/Quick_Sort/Quick_Sort.cpp
#include<iostream>
using namespace std;
void disp(int a[], int n)
{
	for (int i = 0; i < n; i++)
		cout << a[i] << " ";
	cout << endl;
}
int Partition(int a[], int s, int t)
{
	int i = s, j = t;
	int temp = a[s];
	while (i != j)
	{
		while (j > i&&a[j] >= temp)
			j--;
		a[i] = a[j];
		while (i < j&&a[i] <= temp)
			i++;
		a[j] = a[i];
	}
	a[i] = temp;
	return i;
}
void Quick_sort(int a[], int s, int t)
{
	if (s < t)
	{
		int i = Partition(a, s, t);
		Quick_sort(a, s, i);
		Quick_sort(a, i + 1, t);
	}
}
int main()
{
	int n = 10;
	int a[10] = { 12,12,23,45,44,98,1,4,27,88 };
	cout << "before:" << endl;
	disp(a, n);
	cout << "after sort:" << endl;
	Quick_sort(a, 0, n - 1);
	disp(a, n);
	//system("pause");
	return 0;
}

File: 算法设计与分析/Review/Quick_sort.py
#_*_coding:utf-8_*_
#file: Quick_sort.py
def partition(a,s,t):
	i=s
	j=t
	temp=a[i]
	while i!=j:
		while j>i and a[j]>=temp:
			j-=1
		a[i]=a[j]
		while i<j and a[i]<=temp:
			i+=1
		a[j]=a[i]
	a[i]=temp
	return i

def quick_sort(a,s,t):
    if s<t:
        i=partition(a,s,t)
        quick_sort(a,s,i-1)
        quick_sort(a,i+1,t)

def main():
    a = [12, 12, 23, 45, 44, 98, 1, 4, 27, 88]
    quick_sort(a, 0, len(a) - 1)
    print(a)

if __name__=="__main__":
    main()

